bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/ConfigTomlParser.java

```diff
@@ -278,8 +288,34 @@ private static Object retrieveRecordValues(TomlNode tomlNode, String variableNam
             initialValueEntries.put(fieldName, objectValue);
         }
         validateRequiredField(initialValueEntries, recordType, variableName);
-        return ValueCreator
-                .createReadonlyRecordValue(recordType.getPackage(), recordType.getName(), initialValueEntries);
+        if (type.getTag() == TypeTags.RECORD_TYPE_TAG && type.isReadOnly()) {
+            return createReadOnlyFieldRecord(initialValueEntries, recordType, variableName);
+        }
+        return ValueCreator.createReadonlyRecordValue(recordType.getPackage(), recordName, initialValueEntries);
+    }
+
+    private static Object createReadOnlyFieldRecord(Map<String, Object> initialValueEntries, RecordType recordType,
+                                                    String variableName) {
+        for (Map.Entry<String, Field> field : recordType.getFields().entrySet()) {
+            String fieldName = field.getKey();
+            long flag = field.getValue().getFlags();
+            // remove after fixing #28966
+            if (!SymbolFlags.isFlagOn(flag, SymbolFlags.OPTIONAL) && !SymbolFlags.isFlagOn(flag,
```

Comment:
If a record with all `readonly` fields contains a defaultable field and the value is not provided at configuration, we throw an error.